controllers/apply/form-router-next/steps.js

```diff
@@ -132,23 +137,90 @@ module.exports = function(formBuilder) {
+                .filter(fieldName => {
+                    // Retrieve the file from Formidable's parsed data
+                    const uploadedFile = get(req.files, fieldName);
+                    // Ensure a file was actually provided
+                    // (eg. ignore empty file inputs when a file already exists)
+                    return uploadedFile && uploadedFile.size > 0;
+                })
+                .map(fieldName => {
+                    // Log this file as a potential upload
+                    const uploadedFile = get(req.files, fieldName);
+                    return {
+                        fieldName: fieldName,
+                        fileData: uploadedFile
+                    };
+                });
+
+            // Append the file data to the overall form data for validation
+            filesToUpload.forEach(file => {
+                data[file.fieldName] = {
+                    filename: file.fileData.name,
+                    size: file.fileData.size,
+                    type: file.fileData.type
+                };
+            });
 
-            const validationResult = validateForm(form, data);
+            let validationResult = validateForm(form, data);
 
             try {
-                await PendingApplication.saveApplicationState(
-                    currentlyEditingId,
-                    validationResult.value
-                );
-
                 const fieldNamesForStep = flatMap(step.fieldsets, 'fields').map(
                     field => field.name
                 );
 
-                const errorsForStep = validationResult.messages.filter(item =>
+                let errorsForStep = validationResult.messages.filter(item =>
                     fieldNamesForStep.includes(item.param)
                 );
 
+                const [invalidFiles, validFiles] = partition(
+                    filesToUpload,
+                    file => {
+                        return errorsForStep.find(
+                            _ => _.param === file.fieldName
+                        );
+                    }
+                );
+
+                invalidFiles.forEach(file => {
+                    // Remove the (invalid) file information from form data
+                    delete validationResult.value[file.fieldName];
+                });
+
+                // Check if any files were included and handle them (if valid)
+                await Promise.all(
+                    validFiles.map(file =>
+                        s3.uploadFile({
+                            formId: formId,
+                            applicationId: currentlyEditingId,
+                            fileMetadata: file
+                        })
+                    )
+                ).catch(rejection => {
+                    Sentry.captureException(rejection.error);
+                    // Manually create a form error and send the user back to the form
+                    errorsForStep = concat(errorsForStep, {
```

Comment:
Wondering if we can avoid this assignment by directly rendering the step in this catch block.

```js
renderStep(req, res, validationResult.value, [{ 
  msg: `There was an error uploading your file - please try again`,
  param: rejection.fieldName
}]);
```